### PR TITLE
BREAKING: Add state encryption

### LINKED ISSUE #358: Encrypt Snap state
There were requests (such as, but not limited to, #338) to use Snap's `snap_manageState` RPC to store cached private keys. Since this is plain-text storage, the developers intend to encrypt the data in the Snap and store it encrypted.

I don't see a reason not to do it automatically for all the Snaps using the master key.

### PR DESCRIPTION
**Pull request by FrederikBolding**
Fixes #358 

This will be a breaking change.

### CODE REVIEW

File: packages/controllers/src/snaps/SnapController.ts [hunk 1]
```diff
@@ -880,7 +889,25 @@ export class SnapController extends BaseController<
    * @param snapId - The id of the Snap whose state to get.
    */
   async getSnapState(snapId: SnapId): Promise<Json> {
-    return this.state.snapStates[snapId] ?? null;
+    const state = this.state.snapStates[snapId];
+    return state ? this.decryptSnapState(snapId, state) : null;
+  }
+
+  private async getEncryptionKey(snapId: SnapId): Promise<string> {
+    return this._getAppKey(`${snapId}_encryption`);
```
Comment (ritave) [inline packages/controllers/src/snaps/SnapController.ts:897]: Why this is not just `this._getAppKey(snapId);`

**Review by ritave — CHANGES_REQUESTED** [review]
Please add a test with multiple snaps and multiple app keys and check if the encrypted states differ

File: packages/controllers/src/snaps/SnapController.ts (hunk 1 above)
Comment (FrederikBolding) [reply packages/controllers/src/snaps/SnapController.ts:897]: I was hesitant to use a key that is also available to the snap. The snap itself can access the app key for `snapId`, but not for `${snapId}_encryption`.

**Comment by ritave** [comment]
Add a test for corrupted state as well - this will pretty much be the case for all the currently installed Snaps that weren't encrypted

File: packages/controllers/src/snaps/SnapController.ts (hunk 1 above)
Comment (ritave) [reply packages/controllers/src/snaps/SnapController.ts:922]: How can Snap access the app key?

Comment (FrederikBolding) [reply packages/controllers/src/snaps/SnapController.ts:922]: Using `snap_getAppKey`?

**Comment by FrederikBolding** [comment]
> Add a test for corrupted state as well - this will pretty much be the case for all the currently installed Snaps that weren't encrypted

Well, that'll break. That's why it's a breaking change.

**Comment by ritave** [comment]
> > Add a test for corrupted state as well - this will pretty much be the case for all the currently installed Snaps that weren't encrypted
> 
> Well, that'll break. That's why it's a breaking change.

Snap breaking is ok, extension breaking bugs are not ok. Current flask installations can have saved state that will throw an error inside the background process

**Comment by ritave** [comment]
> > > > Add a test for corrupted state as well - this will pretty much be the case for all the currently installed Snaps that weren't encrypted
> > > 
> > > 
> > > Well, that'll break. That's why it's a breaking change.
> > 
> > 
> > Snap breaking is ok, extension breaking bugs are not ok. Current flask installations can have saved state that will throw an error inside the background process
> 
> How will this break the extension? This will throw in the RPC method handler same as many other RPC methods can do?

You're right, it'll be caught there.

**Comment by FrederikBolding** [comment]
This should be merged alongside other breaking changes in the future.

File: packages/controllers/src/snaps/SnapController.ts [hunk 2]
```diff
@@ -367,6 +368,8 @@ type SnapControllerMessenger = RestrictedControllerMessenger<
   AllowedEvents['type']
 >;
 
+type GetAppKey = (subject: string) => Promise<string>;
```
Comment (ritave) [inline packages/controllers/src/snaps/SnapController.ts:371]: I'd prefer we wouldn't prepend `encryption:` to the Snap Id and probably do brittle string operations in the extension.
How about
```typescript
enum AppKeyTypes {
  stateEncryption
}
type GetAppKey = (subject: string, type: AppKeyTypes) => Promise<string>;
```

**Review by ritave — CHANGES_REQUESTED** [review]

**Review by ritave — APPROVED** [review]

**Review by ritave — APPROVED** [review]

**Pull request merged**